virtcontainers/device/manager/utils.go

```diff
@@ -35,3 +40,70 @@ func isVFIO(hostPath string) bool {
 func isBlock(devInfo config.DeviceInfo) bool {
 	return devInfo.DevType == "b"
 }
+
+func IsVFIOLargeBarSpaceDevice(hostPath string) (bool, error) {
+	if !isVFIO(hostPath) {
+		return false, nil
+	}
+	iommuDevicesPath := filepath.Join(config.SysIOMMUPath, filepath.Base(hostPath), "devices")
+	deviceFiles, err := ioutil.ReadDir(iommuDevicesPath)
+	if err != nil {
+		return false, err
+	}
+	// Pass all devices in iommu group
+	for _, deviceFile := range deviceFiles {
+		tokens := strings.Split(deviceFile.Name(), ":")
+		vfioDeviceType := config.VFIODeviceErrorType
+		if len(tokens) == 3 {
+			vfioDeviceType = config.VFIODeviceNormalType
+		} else {
+			tokens = strings.Split(deviceFile.Name(), "-")
+			if len(tokens) == 5 {
+				vfioDeviceType = config.VFIODeviceMediatedType
+			}
+		}
+		deviceFileName := filepath.Join(iommuDevicesPath, deviceFile.Name())
+		var isLarge bool
+		switch vfioDeviceType {
+		case config.VFIODeviceNormalType:
+			if isLarge, err = isLargeBarSpace(filepath.Join(deviceFileName, "resource")); err != nil {
+				return false, err
+			}
+			return isLarge, nil
+		case config.VFIODeviceMediatedType:
+			logrus.Warnf("Check sysfs resource for VFIODeviceMediatedType is not yet supported: %v", deviceFileName)
+		default:
+			logrus.Warnf("Incorrect tokens found while check sysfs resource: %s", deviceFileName)
+		}
+	}
+	return false, nil
+}
+
+func isLargeBarSpace(resourcePath string) (bool, error) {
+	buf, err := ioutil.ReadFile(resourcePath)
+	if err != nil {
+		return false, fmt.Errorf("failed to read sysfs resource: %v", err)
+	}
+	suffix := []string{"", "K", "M", "G", "T"}
+	for rIdx, line := range strings.Split(string(buf), "\n") {
+		cols := strings.Fields(line)
+		start, _ := strconv.ParseUint(cols[0][2:], 16, 64)
+		end, _ := strconv.ParseUint(cols[1][2:], 16, 64)
+		size := end - start + 1
+		sIdx := 0
+		for i := range suffix {
+			if size/1024 < 1 {
+				break
+			}
+			size /= 1024
+			sIdx = i + 1
+		}
+		logrus.WithField("resource", resourcePath).Infof("Region %d: start=%s(%d) end=%s(%d) size:%d%s",
```

Comment:
`WithFields` this can look better

Reply:
Need a formatted output here:)

Reply:
ooi why is that? Is something post-processing this log message?

Reply:
updated.